gazelle/generate.go

```diff
@@ -0,0 +1,309 @@
+// GenerateRules is called in each directory where an update is requested
+// in depth-first post-order.
+func (ts *TypeScript) GenerateRules(args language.GenerateArgs) language.GenerateResult {
+	cfgs := args.Config.Exts[languageName].(Configs)
+	cfg := cfgs[args.Rel]
+
+	// When we return empty, we mean that we don't generate anything, but this
+	// still triggers the indexing for all the TypeScript targets in this
+	// package.
+	if !cfg.GenerationEnabled() {
+		return language.GenerateResult{}
+	}
+
+	// If this directory has not been declared as a bazel package it will have been
+	// including in the parent BUILD file.
+	if !isBazelPackage(args.Dir) {
+		return language.GenerateResult{}
+	}
+
+	// Collect all source files
+	sourceFiles, dataFiles, collectErr := collectSourceFiles(cfg, args)
+	if collectErr != nil {
+		log.Printf("ERROR: %v\n", collectErr)
+		return language.GenerateResult{}
+	}
+
+	DEBUG("SOURCE(%q): %s", args.Rel, sourceFiles.Values())
+
+	// Divide src vs test files
+	libSourceFiles := treeset.NewWithStringComparator()
+	testSourceFiles := treeset.NewWithStringComparator()
+
+	for _, f := range sourceFiles.Values() {
+		file := f.(string)
+		if cfg.IsTestFile(file) {
+			testSourceFiles.Add(file)
+		} else if cfg.IsSourceFile(file) {
+			libSourceFiles.Add(file)
+		}
+	}
+
+	// Build the GenerateResult with src and test rules
+	var result language.GenerateResult
+
+	addProjectRule(
+		cfg,
+		args,
+		cfg.RenderLibraryName(filepath.Base(args.Dir)),
+		libSourceFiles,
+		dataFiles,
+		&result,
+	)
+
+	addProjectRule(
+		cfg,
+		args,
+		cfg.RenderTestsLibraryName(filepath.Base(args.Dir)),
+		testSourceFiles,
+		dataFiles,
+		&result,
+	)
+
+	return result
+}
+
+func addProjectRule(cfg *TypeScriptConfig, args language.GenerateArgs, targetName string, sourceFiles, dataFiles *treeset.Set, result *language.GenerateResult) {
+	// Generate nothing if there are no source files
+	if sourceFiles.Empty() {
+		// If there is already a BUILD then potentially clean it up
+		if args.File != nil {
+			// Remove any exiting instance of this rule project rule
+			for _, r := range args.File.Rules {
+				if r.Name() == targetName && r.Kind() == tsProjectKind {
+					result.Empty = append(result.Empty, rule.NewRule(r.Kind(), r.Name()))
+				}
+			}
+		}
+
+		return
+	}
+
+	// If a build already exists check for name-collisions with the rule being generated
+	if args.File != nil {
+		checkCollisionErrors(targetName, args)
+	}
+
+	// Data files imported by sourceFiles
+	sourceDataFiles := treeset.NewWithStringComparator()
+
+	// Collect import statements from source
+	importedFiles := treeset.NewWith(importStatementComparator)
+
+	// TODO(jbedard): parse files concurrently
+	sourceFileIt := sourceFiles.Iterator()
+	for sourceFileIt.Next() {
+		filePath := sourceFileIt.Value().(string)
+		fileImports, err := parseFile(filepath.Join(args.Dir, filePath))
+
+		if err != nil {
+			fmt.Println("Parse Error:", fmt.Errorf("%q: %v", filePath, err))
```

Comment:
A parsing error is non-fatal?

Reply:
We don't want to to `os.Exit` just because of a single file with bad syntax, just output an error/warning and continue generating BUILDs